### PR TITLE
cgroup2: devices filtering cleanup

### PR DESCRIPTION
**Pull request by cyphar**
This is my attempt at solving both #2797 and #2366. I'm not entirely sure if the behaviour I've implemented if a cgroup has more than one filter installed (detach all old filters, so that only the new one is effective) is quite correct, but it's necessary to fix the program leak problem (outside of switching to BPF pins entirely).

NOTE: I wanted to add some unit tests for our devicefilter code, but unfortunately `BPF_PROG_TEST_RUN` doesn't support cgroup programs at the moment. I would have to push some code upstream for that to happen (though I'll look into that next week).

<hr>

Changelog entry:

```
cgroup2: devices: rework the filter generation to produce consistent results with cgroupv1, and always clobber any existing eBPF program(s) to fix `runc update` and avoid leaking eBPF programs (resulting in errors when managing containers). 
```

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

### CODE REVIEW

File: libcontainer/cgroups/ebpf/devicefilter/devicefilter.go [hunk 1]
```diff
@@ -10,8 +10,10 @@ import (
 	"math"
 	"strconv"
 
-	"github.com/cilium/ebpf/asm"
+	devicesemulator "github.com/opencontainers/runc/libcontainer/cgroups/devices"
 	"github.com/opencontainers/runc/libcontainer/devices"
+
```
Comment (kolyshkin) [inline libcontainer/cgroups/ebpf/devicefilter/devicefilter.go:15]: Remove the extra newline.

Or, if you want to switch to 3 groups of imports, the order should be
1. standard
2. external third-party
3. "local" (as in `goimports --local`)

IOW `github.com/opencontainers/runc/` should be the last group.

File: libcontainer/cgroups/ebpf/devicefilter/devicefilter.go [hunk 2]
```diff
@@ -68,38 +103,32 @@ func (p *program) init() {
 }
 
 // appendDevice needs to be called from the last element of OCI linux.resources.devices to the head element.
```
Comment (kolyshkin) [inline libcontainer/cgroups/ebpf/devicefilter/devicefilter.go:105]: `s/appendDevice/appendRule/`

Comment (kolyshkin) [reply libcontainer/cgroups/ebpf/devicefilter/devicefilter.go:105]: Also seems that this entire comment is now obsoleted.

**Comment by kolyshkin** [comment]
A typo in the second commit message: `s/speciifc/specific/`.

File: libcontainer/cgroups/ebpf/ebpf.go [hunk 3]
```diff
@@ -1,29 +1,93 @@
...
+		QueryType   uint32
+		AttachFlags uint32
+		ProgIds     uint64 // __aligned_u64
+		ProgCnt     uint32
+	}
+
+	// Currently you can only have 64 eBPF programs attached to a cgroup.
+	size := 64
+	allowRetry := true
+	for allowRetry {
```
Comment (kolyshkin) [inline libcontainer/cgroups/ebpf/ebpf.go:32]: It seems that `allowRetry` is redundant; you can just do

```go
for {
	...
	if errno != 0 {
		// On ENOSPC we get the correct number of programs.
		if errno == unix.ENOSPC {
			size = int(query.ProgCnt)
			continue
		}
		return nil, fmt.Errorf("bpf_prog_query: %w", errno)
	}
	...
	return ...
}
```

File: libcontainer/cgroups/ebpf/ebpf.go [hunk 4]
```diff
@@ -1,29 +1,93 @@
...
+			// On ENOSPC we get the correct number of programs.
+			if errno == unix.ENOSPC {
+				allowRetry = false
+				continue
+			}
+			return nil, errno
+		}
+
+		// Convert the ids to program handles.
+		programs := make([]*ebpf.Program, len(progIds))
```
Comment (kolyshkin) [inline libcontainer/cgroups/ebpf/ebpf.go:58]: should this be `size` rather than `len(progIds)`?

File: libcontainer/cgroups/ebpf/ebpf.go [hunk 5]
```diff
@@ -1,29 +1,93 @@
...
+		for idx, progId := range progIds {
```
Comment (kolyshkin) [inline libcontainer/cgroups/ebpf/ebpf.go:59]: same here: `range progIds[:size]`. Or have `progIds = progIds[:size]` above (line 47).

File: libcontainer/cgroups/ebpf/ebpf.go [hunk 6]
```diff
@@ -1,29 +1,93 @@
...
+			if err != nil {
+				return nil, errors.Wrap(err, "cannot fetch program from id")
+			}
+			programs[idx] = program
+		}
+		runtime.KeepAlive(progIds)
+		return programs, nil
+	}
+
+	return nil, errors.Errorf("could not get list of CGROUP_DEVICE programs")
```
Comment (kolyshkin) [inline libcontainer/cgroups/ebpf/ebpf.go:70]: nit: use `errors.New`

File: libcontainer/cgroups/ebpf/ebpf.go [hunk 7]
```diff
@@ -1,29 +1,93 @@
...
+			}
+			return nil, errno
+		}
+
+		// Convert the ids to program handles.
+		programs := make([]*ebpf.Program, len(progIds))
+		for idx, progId := range progIds {
+			program, err := ebpf.NewProgramFromID(ebpf.ProgramID(progId))
+			if err != nil {
+				return nil, errors.Wrap(err, "cannot fetch program from id")
```
Comment (kolyshkin) [inline libcontainer/cgroups/ebpf/ebpf.go:62]: The new golang style to wrap an error is `fmt.Errorf("cannot fetch program from id: %w", err)`, but we still have a lot of `errors.Wrap()` so whatever.

File: libcontainer/cgroups/ebpf/devicefilter/devicefilter.go (hunk 1 above)
Comment (cyphar) [reply libcontainer/cgroups/ebpf/devicefilter/devicefilter.go:15]: Interesting, I've always done external third-party last (so it's in order of "locality"). But sure, I can drop the newline.

File: libcontainer/cgroups/ebpf/ebpf.go (hunk 3 above)
Comment (cyphar) [reply libcontainer/cgroups/ebpf/ebpf.go:32]: I don't want to loop endlessly ... but you're quite right that the `allowRetry` logic was wrong (oops). I'll fix it up.

File: libcontainer/cgroups/ebpf/ebpf.go (hunk 4 above)
Comment (cyphar) [reply libcontainer/cgroups/ebpf/ebpf.go:58]: Yup, you're quite right.

File: libcontainer/cgroups/ebpf/ebpf.go (hunk 7 above)
Comment (cyphar) [reply libcontainer/cgroups/ebpf/ebpf.go:62]: I'll switch them all to `Errorf("%w")`.

Comment (AkihiroSuda) [reply libcontainer/cgroups/ebpf/ebpf.go]: What is the problem with `errors.Wrap`?

Comment (cyphar) [reply libcontainer/cgroups/ebpf/ebpf.go]: There's nothing wrong with `errors.Wrap` specifically, but the newer `fmt.Errorf("%w")` setup is in the stdlib and the stdlib `errors.As` and `errors.Is` are probably going to end up more widely used (`pkg/errors` does still have the upside that you get a stack trace though...).

**Comment by cyphar** [comment]
We also can't test the `runc update` behaviour because it turns out `runc update` doesn't support cgroup device limits... I tried adding support for it, but the behaviour is a bit weird because of the default allowed devices... Hmmm....

**Comment by kolyshkin** [comment]
> We also can't test the runc update behaviour because it turns out runc update doesn't support cgroup device limits... I tried adding support for it, but the behaviour is a bit weird because of the default allowed devices... Hmmm....

Yeah, `update.go` basically repeats specconv, a year ago I tried to reuse the specconv in update but got nowhere...

Maybe we can at least have a simple test case added to libcontainer/integration, which removes and re-adds a device allow rule say 100 times (and check that device access works as expected). Without this PR we should fail on 64th iteration due to eBPF progs limit reached.

Same for runc update -- we could have a stupid test calling `runc update` 100 times (without actually changing anything) but I am not sure if such test makes sense if we implement what's described in the previous paragraph.

**Review by kolyshkin — APPROVED** [review]
LGTM except for the missing integration test(s).

**Comment by cyphar** [comment]
Yeah looking at it I think we would need to rethink the "default allow" device list behaviour if we ever want to add devices support to `runc update` (otherwise users will accidentally brick their containers when using it). I'll cook up a unit test on Monday.

**Comment by cyphar** [comment]
Seems as though the Microsoft PPA is down?

File: tests/integration/update.bats [hunk 8]
```diff
@@ -648,3 +648,30 @@ EOF
...
+	# In the past runc would not delete old cgroupv2 eBPF programs, so this
+	# test ensures that once we go past the program limit (64 stacked programs
+	# at time of writing) you can still run "runc" update.
+
+	# Run the container in the background.
+	runc run -d --console-socket "$CONSOLE_SOCKET" test_update
+	[ "$status" -eq 0 ]
+
+	for new_limit in $(seq 300)
+	do
```
Comment (kolyshkin) [inline tests/integration/update.bats:670]: `make shfmt` suggests:
```diff
-	for new_limit in $(seq 300)
-	do
+	for new_limit in $(seq 300); do
```

I like it either way but having a single coding style is good.

**Comment by kolyshkin** [comment]
CI restarted (looks like it's working today).

**Comment by kolyshkin** [comment]
CI restarted; GHA still has some issues :(

> E: Failed to fetch https://packages.microsoft.com/ubuntu/20.04/prod/dists/focal/main/binary-amd64/Packages.bz2  File has unexpected size (73402 != 73160). Mirror sync in progress? [IP: 104.42.185.173 443]

**Comment by cyphar** [comment]
CI passes again now. :smile_cat:

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by kolyshkin** [comment]
@AkihiroSuda @mrunalp PTAL

File: tests/integration/update.bats [hunk 9]
```diff
@@ -648,3 +648,29 @@ EOF
...
+	#
+	# In the past runc would not delete old cgroupv2 eBPF programs, so this
+	# test ensures that once we go past the program limit (64 stacked programs
+	# at time of writing) you can still run "runc" update.
+
+	# Run the container in the background.
+	runc run -d --console-socket "$CONSOLE_SOCKET" test_update
+	[ "$status" -eq 0 ]
+
+	for new_limit in $(seq 300); do
```
Comment (AkihiroSuda) [inline tests/integration/update.bats:669]: seq might cause extra alloc compared to `for ((i = 0; i < 300; i++))`, but negligible

File: tests/integration/update.bats [hunk 10]
```diff
@@ -648,3 +648,29 @@ EOF
 	runc resume test_update
 	[ "$status" -eq 0 ]
 }
+
+@test "runc update replaces devices cgroup program" {
+	[[ "$ROOTLESS" -ne 0 ]] && requires rootless_cgroup
```
Comment (AkihiroSuda) [inline tests/integration/update.bats:653]: The test is NOP in rootless, right? Can we have a comment line to explain the expected behavior?

File: libcontainer/cgroups/ebpf/ebpf.go [hunk 11]
```diff
@@ -1,29 +1,95 @@
...
+			unsafe.Sizeof(query))
+		size = int(query.ProgCnt)
+		runtime.KeepAlive(query)
+		if errno != 0 {
+			// On ENOSPC we get the correct number of programs.
+			if errno == unix.ENOSPC {
+				retries++
+				continue
+			}
+			return nil, fmt.Errorf("bpf_prog_query(BPF_CGROUP_DEVICE) failed: %w", errno)
```
Comment (AkihiroSuda) [inline libcontainer/cgroups/ebpf/ebpf.go:55]: We may get EINTR wand want to retry?

File: libcontainer/integration/exec_test.go [hunk 12]
```diff
@@ -1921,6 +1921,9 @@ func TestFdLeaks(t *testing.T) {
 }
 
 func TestFdLeaksSystemd(t *testing.T) {
+	if !systemd.IsRunningSystemd() {
+		t.Skip("Systemd is unsupported")
```
Comment (AkihiroSuda) [inline libcontainer/integration/exec_test.go:1925]: "test requires systemd" might be less confusing.

Comment (kolyshkin) [reply libcontainer/integration/exec_test.go:1925]: @AkihiroSuda this is what we already have in other cases:

```
--
libcontainer/integration/exec_test.go:  if !systemd.IsRunningSystemd() {
libcontainer/integration/exec_test.go-          t.Skip("Systemd is unsupported")
--
libcontainer/integration/exec_test.go:  if !systemd.IsRunningSystemd() {
libcontainer/integration/exec_test.go-          t.Skip("Systemd is unsupported")
--
libcontainer/integration/exec_test.go:  if !systemd.IsRunningSystemd() {
libcontainer/integration/exec_test.go-          t.Skip("Systemd is unsupported")
--
libcontainer/integration/exec_test.go:  if !systemd.IsRunningSystemd() {
libcontainer/integration/exec_test.go-          t.Skip("Systemd is unsupported")
--
libcontainer/integration/exec_test.go:  if !systemd.IsRunningSystemd() {
libcontainer/integration/exec_test.go-          t.Skip("Systemd is unsupported")
--
```

File: libcontainer/cgroups/ebpf/ebpf.go (hunk 11 above)
Comment (kolyshkin) [reply libcontainer/cgroups/ebpf/ebpf.go:55]: bpf(2) is considered EINTR-safe. Can not quote any sources though :(

Comment (cyphar) [reply libcontainer/cgroups/ebpf/ebpf.go:55]: Seems that you can get EINTR from `bpf` (especially in Go thanks to green thread preemption). cilium/ebpf#24

File: libcontainer/integration/exec_test.go (hunk 12 above)
Comment (cyphar) [reply libcontainer/integration/exec_test.go:1925]: Yeah I copied this from the other integration functions. If we really want to change it we can but we should change it for all of them in one commit.

File: tests/integration/update.bats (hunk 10 above)
Comment (cyphar) [reply tests/integration/update.bats:653]: It's also no-op for cgroupv1...

File: libcontainer/cgroups/ebpf/devicefilter/devicefilter.go [hunk 13]
```diff
@@ -11,6 +11,7 @@ import (
 	"strconv"
 
 	"github.com/cilium/ebpf/asm"
+	devicesemulator "github.com/opencontainers/runc/libcontainer/cgroups/devices"
```
Comment (kolyshkin) [inline libcontainer/cgroups/ebpf/devicefilter/devicefilter.go:14]: nit: `devicesEmulator`

File: libcontainer/cgroups/ebpf/ebpf.go (hunk 11 above)
Comment (kolyshkin) [reply libcontainer/cgroups/ebpf/ebpf.go:55]: Found another case -- they retry on `bpf(BPF_PROG_LOAD)` but on `EAGAIN` not `EINTR`.

So I guess we don't need a EINTR handler here.

**Review by kolyshkin — APPROVED** [review]
LGTM

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**